bootstrap/kubeadm/api/equality/semantic.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package equality
```

Comment:
Are these only used for adoption path? I'm wondering if we should put them under the internal package and drop them in v1alpha4

Reply:
Right now, yes.

The question is: what are we going to do for v1a4 to avoid dropping a new KCP controller next to an existing v1a3 control plane and having it go on and "upgrade" every CP machine it sees until the hashes match?

The way the deployment controller solves this problem upstream is by eschewing the hash for comparison and using a "semantic equals" on the pod templates. That's the route I imagined taking with this package.

Reply:
I'm not 100% if i'm following, but I was hoping we could drop adoption when v1alpha4 is released and have one true way to spin up a control plane for cluster api

Reply:
Ah, this is both at 90 degrees to the change and a longer conversation. Even w/o adoption we'd still want to consider the semantic equals thing to make it possible to change the KCP type in the future, and I have some examples about why adoption is useful outside of migration contexts.